src/order/monotone.lean

```diff
@@ -0,0 +1,626 @@
+Authors: Jeremy Avigad, Mario Carneiro, Yaël Dillies
+-/
+import order.basic
+
+/-!
+# Monotonicity
+
+This file defines (strictly) monotone/antitone functions. Contrary to standard mathematical usage,
+"monotone"/"mono" here means "increasing", not "increasing or decreasing". We use "antitone"/"anti"
+to mean "decreasing".
+
+## Definitions
+
+* `monotone f`: A function `f` between two preorders is monotone if `a ≤ b` implies `f a ≤ f b`.
+* `antitone f`: A function `f` between two preorders is antitone if `a ≤ b` implies `f b ≤ f a`.
+* `monotone_on f s`: Same as `monotone f`, but for all `a, b ∈ s`.
+* `antitone_on f s`: Same as `antitone f`, but for all `a, b ∈ s`.
+* `strict_mono f` : A function `f` between two preorders is strictly monotone if `a < b` implies
+  `f a < f b`.
+* `strict_anti f` : A function `f` between two preorders is strictly antitone if `a < b` implies
+  `f b < f a`.
+* `strict_mono_on f s`: Same as `strict_mono f`, but for all `a, b ∈ s`.
+* `strict_anti_on f s`: Same as `strict_anti f`, but for all `a, b ∈ s`.
+
+## Main theorems
+
+* `monotone_nat_of_le_succ`: If `f : ℕ → α` and `f n ≤ f (n + 1)` for all `n`, then `f` is
+  monotone.
+* `antitone_nat_of_succ_le`: If `f : ℕ → α` and `f (n + 1) ≤ f n` for all `n`, then `f` is
+  antitone.
+* `strict_mono_nat_of_lt_succ`: If `f : ℕ → α` and `f n < f (n + 1)` for all `n`, then `f` is
+  strictly monotone.
+* `strict_anti_nat_of_succ_lt`: If `f : ℕ → α` and `f (n + 1) < f n` for all `n`, then `f` is
+  strictly antitone.
+
+## Implementation notes
+
+Some of these definitions used to only require `has_le α` or `has_lt α`. The advantage of this is
+unclear and it led to slight elaboration issues. Now, everything requires `preorder α` and seems to
+work fine. Related Zulip discussion:
+https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Order.20diamond/near/254353352.
+
+## TODO
+
+The above theorems are also true in `ℤ`, `ℕ+`, `fin n`... To make that work, we need `succ_order α`
+along with another typeclass we don't yet have roughly stating "everything is reachable using
+finitely many `succ`".
+
+## Tags
+
+monotone, strictly monotone, antitone, strictly antitone, increasing, strictly increasing,
+decreasing, strictly decreasing
+-/
+
+open function
+
+universes u v w
+variables {α : Type u} {β : Type v} {γ : Type w} {r : α → α → Prop}
+
+section monotone_def
+variables [preorder α] [preorder β]
+
+/-- A function `f` is monotone if `a ≤ b` implies `f a ≤ f b`. -/
+def monotone (f : α → β) : Prop := ∀ ⦃a b⦄, a ≤ b → f a ≤ f b
+
+/-- A function `f` is antitone if `a ≤ b` implies `f b ≤ f a`. -/
+def antitone (f : α → β) : Prop := ∀ ⦃a b⦄, a ≤ b → f b ≤ f a
+
+/-- A function `f` is antitone on `s` if, for all `a, b ∈ s`, `a ≤ b` implies `f b ≤ f a`. -/
+def monotone_on (f : α → β) (s : set α) : Prop :=
+∀ ⦃a⦄ (ha : a ∈ s) ⦃b⦄ (hb : b ∈ s), a ≤ b → f a ≤ f b
+
+/-- A function `f` is monotone on `s` if, for all `a, b ∈ s`, `a ≤ b` implies `f a ≤ f b`. -/
```

Comment:
```suggestion
/-- A function `f` is monotone on `s` if, for all `a, b ∈ s`, `a ≤ b` implies `f b ≤ f a`. -/
def monotone_on (f : α → β) (s : set α) : Prop :=
∀ ⦃a⦄ (ha : a ∈ s) ⦃b⦄ (hb : b ∈ s), a ≤ b → f a ≤ f b

/-- A function `f` is antitone on `s` if, for all `a, b ∈ s`, `a ≤ b` implies `f a ≤ f b`. -/
```

Reply:
Oh what I got them wrong originally. Thanks for spotting!